### PR TITLE
heFFTe v1.0.1 API update

### PR DESCRIPTION
**Pull request by af-ayala**
Integration of heFFTe v 1.0.1 into Cabana ECP.
The current API (fully based on C++11) shall remain for future developments.

### CODE REVIEW

**Review by streeve — DISMISSED** [review]
Thanks for putting this together. Looks like travis is failing because of formatting (we enforce clang-format 7)

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 1]
```diff
@@ -52,101 +52,102 @@ struct HeffteMemoryTraits<Kokkos::HostSpace>
...
+//     // Buffer packing type.
+//     int packflag;
+
+//     // Forward scaling option.
+//     int scaled;
+// };
 
 //---------------------------------------------------------------------------//
-template <class Scalar, class EntityType, class MeshType, class DeviceType>
+template <class Backend, class Scalar, class EntityType, class MeshType, class DeviceType>
```
[the remark this answers is not in the document]
Comment (sfogerty) [reply cajita/src/Cajita_FastFourierTransform.hpp:150]: I think so - seems wise. Should we try to incorporate that restriction now, or kick that can down the road a bit?

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 2]
```diff
@@ -163,25 +164,26 @@ class FastFourierTransform
...
+        // : _fft( layout.localGrid()->globalGrid().comm() )
     {
         if ( 1 != layout.dofsPerEntity() )
             throw std::logic_error(
                 "Only 1 complex value per entity allowed in FFT" );
 
         // Set the memory type. For now we will just do FFTs on the host until
         // we find the HEFFTE GPU memory bug.
-        _fft.mem_type = HEFFTE_MEM_CPU;
+        // _fft.mem_type = HEFFTE_MEM_CPU;  // * Memory automatically defined from backend type
```
Comment (sfogerty) [inline cajita/src/Cajita_FastFourierTransform.hpp:177]: This implies that the `HeffteMemoryTraits` (lines 31-52) don’t seem to be needed anymore. So I think we should go ahead and remove them.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 3]
```diff
@@ -214,21 +216,35 @@ class FastFourierTransform
...
         if ( fftsize < (int)entity_space.size() )
-            throw std::logic_error( "HEFFTE expected allocation size smaller "
+            throw std::logic_error( "heFFTe expected allocation size smaller "
                                     "than local grid size" );
 
         // Allocate the work array.
         _fft_work = Kokkos::View<Scalar *, DeviceType>(
             Kokkos::ViewAllocateWithoutInitializing( "fft_work" ),
-            2 * fftsize );
+            2 * fftsize );
```
[the remark this answers is not in the document]
Comment (sfogerty) [reply cajita/src/Cajita_FastFourierTransform.hpp:242]: I'm somewhat confused here...  What I understand is you could use just allocate `fftsize` if defining input and output data to be complex? This would be for a complex-to-complex, in-place FFT. 

Could we do something like:
 ```
 _fft_work = Kokkos::View<Kokkos::complex<Scalar> *, DeviceType>(
            Kokkos::ViewAllocateWithoutInitializing( "fft_work" ),
            fftsize );

```

**Comment by streeve** [comment]
Sorry, @junghans I broke it again

**Comment by sfogerty** [comment]
@streeve @junghans My reasoning to build with `Cabana_ENABLE_HEFFTE` and not `Cabana_REQUIRE_HEFFTE` is in the details of `Cabana_add_dependency()`. 
If we set the `REQUIRE` var to ON, then we are calling `find_package(Heffte)` twice. Unfortunately Heffte v1.0 is not guarded against this, and a CMake error occurs. (My Heffte [PR](https://bitbucket.org/icl/heffte/pull-requests/83/guard-for-multiple-targets-of-same-lib/diff) just a couple commits after v1.0)

**Comment by streeve** [comment]
`Cabana_ENABLE_HEFFTE` doesn't do anything, but I understand why you removed `REQUIRE` now. We'll need a minor release for that then

**Comment by sfogerty** [comment]
> `Cabana_ENABLE_HEFFTE` doesn't do anything, but I understand why you removed `REQUIRE` now. We'll need a minor release for that then

Ah, right. I'm not sure there is an ideal solution at the moment. We could, for now, have travis just clone and build with the most recent heFFTe commit (not glued to v1.0 tag). Our build will start breaking when a new heFFTe version releases, but at that point we can redirect travis to build that new tagged version (v1.0.1, or whatever)?

File: CMakeLists.txt [hunk 4]
```diff
@@ -50,14 +50,15 @@ endif()
...
+  cmake_parse_arguments(CABANA_DEPENDENCY "" "PACKAGE;VERSION" "" ${ARGN})
   string(TOUPPER "${CABANA_DEPENDENCY_PACKAGE}" CABANA_DEPENDENCY_OPTION )
   option(
     Cabana_REQUIRE_${CABANA_DEPENDENCY_OPTION}
     "Require Cabana to build with ${CABANA_DEPENDENCY_PACKAGE} support" ${CABANA_DEPENDENCY_PACKAGE}_FOUND)
   if(Cabana_REQUIRE_${CABANA_DEPENDENCY_OPTION})
-    find_package( ${CABANA_DEPENDENCY_PACKAGE} REQUIRED )
+    find_package( ${CABANA_DEPENDENCY_PACKAGE} ${CABANA_DEPENDENCY_VERSION} REQUIRED )
+  else()
+    find_package( ${CABANA_DEPENDENCY_PACKAGE} ${CABANA_DEPENDENCY_VERSION} QUIET )
```
Comment (junghans) [inline CMakeLists.txt:61]: You cannot rearrange this as the result (`*_FOUND`) is used in the `option()` call.

**Review by junghans — DISMISSED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (junghans) [reply CMakeLists.txt:61]: I would say put it back for now, and just call `find_package(Heffte 1.0.0)` below.

Comment (sfogerty) [reply CMakeLists.txt]: Reverted to original, and using the new heFFTe v1.0.1 now 👍

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 5]
```diff
@@ -28,232 +28,180 @@ namespace Cajita
...
+            Kokkos::ViewAllocateWithoutInitializing( "fft_work" ), fftsize );
     }
 
     /*!
       \brief Do a forward FFT.
       \param in The array on which to perform the forward transform.
     */
     template <class Array_t>
-    void forward( const Array_t &x )
+    void forward( const Array_t &x, const FFTScaleNone )
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:168]: These should be an `override` implementation of a pure virtual function in the base class. That way future FFT implementations have to conform to the same interface.

**Review by sslattery — CHANGES_REQUESTED** [review]

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 5 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp:168]: That was my plan, but how would I do that with the template?

**Comment by streeve** [comment]
This should be ready for another review

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 5 above)
Comment (sslattery) [reply cajita/src/Cajita_FastFourierTransform.hpp:219]: explicit overloads with each template variant

**Comment by sslattery** [comment]
You guys will need to do a rebase to resolve conflicts - they are likely from recent updates to the code formatting style.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 6]
```diff
@@ -28,232 +28,281 @@ namespace Cajita
...
+        compute( x, -1, heffte::scale::symmetric );
+    }
+
+    template <class ComplexType>
+    KOKKOS_INLINE_FUNCTION ComplexType copyFromKokkosComplex(
+        Kokkos::complex<value_type> x_view_val, ComplexType work_view_val,
+        typename std::enable_if<( is_cuda_complex<ComplexType>::value ),
+                                int>::type * = 0 )
+    {
+        auto realpart = x_view_val.real();
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:260]: why make the temporaries here?

**Review by sslattery — CHANGES_REQUESTED** [review]
Other than my few minor comments this looks great!

**Comment by sfogerty** [comment]
Retest this please

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 6 above)
Comment (sfogerty) [reply cajita/src/Cajita_FastFourierTransform.hpp]: No need really - I removed them

File: cajita/unit_test/tstFastFourierTransform.hpp [hunk 7]
```diff
@@ -31,7 +31,8 @@ using namespace Cajita;
 namespace Test
 {
 //---------------------------------------------------------------------------//
-void memoryTest()
+// New heFFTe version - this should be re-done
+/*void memoryTest()
```
[the remark this answers is not in the document]
Comment (sfogerty) [reply cajita/unit_test/tstFastFourierTransform.hpp]: Removed this test

**Comment by streeve** [comment]
I rebased to clean up the commit history. 

Also just a note that the interpolation test failed again (2 ranks)

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 8]
```diff
@@ -28,232 +28,281 @@ namespace Cajita
...
       \param in The array on which to perform the forward transform.
     */
     template <class Array_t>
-    void forward( const Array_t &x )
+    void forward( const Array_t &x, const FFTScaleNone )
+    {
+        compute( x, 1, heffte::scale::none );
+    }
+    template <class Array_t>
+    void forward( const Array_t &x, const FFTScaleFull )
```
[the remark this answers is not in the document]
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Done

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 5 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: At your suggestion I used CRTP for this. It's a much better design, but I couldn't see how to simplify the template parameters of the derived class like you mentioned.

**Comment by streeve** [comment]
@junghans any idea why codecov is messed up?

**Comment by junghans** [comment]
Try to merge with master first.

File: cajita/unit_test/tstFastFourierTransform.hpp [hunk 9]
```diff
@@ -97,19 +86,53 @@ void forwardReverseTest()
...
+    }
+    else if ( use_default )
+    {
+        auto fft =
+            Experimental::createHeffteFastFourierTransform<double, TEST_DEVICE>(
+                *vector_layout );
+        fft->forward( *lhs, Experimental::FFTScaleFull() );
+        fft->reverse( *lhs, Experimental::FFTScaleNone() );
+    }
+#ifndef TEST_CUDA
```
Comment (sslattery) [inline cajita/unit_test/tstFastFourierTransform.hpp:120]: Why is this here?

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 10]
```diff
@@ -23,291 +23,509 @@
 #include <memory>
 #include <type_traits>
 
+#ifdef KOKKOS_ENABLE_HIP // FIXME_HIP
+#error FFT not supported with Kokkos::Experimental::HIP backend
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:27]: This error should be a `static_assert` in the HIP specialization instead of a macro-level error here. That way it would only fail when people actually tried to use it. The way this is written here it will possibly preclude users from having HIP kernels and an FFT on the host.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 11]
```diff
@@ -23,291 +23,509 @@
...
+*/
+struct FFTScaleSymmetric
+{
+};
+
+/*!
+  \class FFTBackendFFTW
+  \brief Tag specifying FFTW backend for FFT (host default).
+*/
+struct FFTBackendFFTW
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:62]: These backend options are heFFTe specific. All backend stuff should be put in some kind of heFFTe namespace or in the heFFTe class.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 12]
```diff
@@ -23,291 +23,509 @@
...
-      0: point-to-point
-      1: all-to-all
-      2: combination
+//---------------------------------------------------------------------------//
+/*!
+  \class FastFourierTransform
+  \brief 3D distributed fast Fourier transform base implementation.
+*/
+template <class EntityType, class MeshType, class Scalar, class DeviceType,
+          class InterfaceType>
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:167]: Rename `InterfaceType` to `Derived`. It will make it very obvious that you implemented the CRTP.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 13]
```diff
@@ -23,291 +23,509 @@
...
+      \param x The array on which to perform the forward transform.
+      \param scaling Method of scaling data.
     */
-    FastFourierTransformParams& setPackType( const int type )
+    template <class Array_t, class ScaleType>
+    void forward( const Array_t& x, const ScaleType scaling )
     {
-        packflag = type;
-        return *this;
+        checkArray( x );
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:248]: A better way to do this would be to write some kind of meta type checker (i.e a concept) instead of having a function which only does compile-time work. You could simply make all those `is_*` checks in the `static_assert` calls another type checker that would inherit from `std::is_true` only if they are all satisfied.

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 11 above)
Comment (sslattery) [reply cajita/src/Cajita_FastFourierTransform.hpp:62]: So I would put them in the `Impl` namespace below and prefix all the struct names with `Heffte`.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 14]
```diff
@@ -23,291 +23,509 @@
...
+#ifdef Heffte_ENABLE_CUDA
+#ifdef KOKKOS_ENABLE_CUDA
+template <>
+struct HeffteBackendTraits<Kokkos::Cuda, double, Impl::FFTBackendDefault>
+{
+    using backend_type = heffte::backend::cufft;
+    using complex_type = cufftDoubleComplex;
+};
+template <>
+struct HeffteBackendTraits<Kokkos::Cuda, float, Impl::FFTBackendDefault>
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:306]: Here is where I would write a HIP specialization and just put `static_assert( false, "HIP not supported" );`

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 11 above)
Comment (sslattery) [reply cajita/src/Cajita_FastFourierTransform.hpp:62]: If we find that other FFT libraries do the same thing we can relax this and generalize the backend concept to some kind of 1D FFT concept.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 15]
```diff
@@ -23,291 +23,509 @@
...
-        if ( 1 != x.layout()->dofsPerEntity() )
-            throw std::logic_error(
-                "Only 1 complex value per entity allowed in FFT" );
+    /*!
+     \brief Copy data from Kokkos::complex to std::complex.
+     \param x_view_val Kokkos::complex value.
+     \param work_view_val std::complex value.
+    */
+    template <class ComplexType>
+    ComplexType copyFromKokkosComplex(
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:459]: These need to be `KOKKOS_INLINE_FUNCTION`. Even if `std::complex` won't run on device we want it to be `inline`.

**Review by sslattery — CHANGES_REQUESTED** [review]

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 10 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Done

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 11 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp:58]: I left them for now because the same tags could be used for the other FFT libraries. The only issue there is disabling backends that don't exist for a given library (fftMPI has one more option than heFFTe)

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 12 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Done

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 13 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Done

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 15 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: I used `inline` since it can't be called from the device

File: cajita/unit_test/tstFastFourierTransform.hpp (hunk 9 above)
Comment (streeve) [reply cajita/unit_test/tstFastFourierTransform.hpp:120]: This section only tests the explicit use of an FFT backend, rather than the default for the Kokkos backend. Since there are only options for the host, I needed to disable this from the automatic test expansion on all Kokkos backends

**Comment by streeve** [comment]
Retest this please

**Comment by streeve** [comment]
This is ready for review (HIP failure is unrelated - not yet using HIP with FFT)

File: cajita/unit_test/tstFastFourierTransform.hpp [hunk 16]
```diff
@@ -30,21 +30,10 @@ using namespace Cajita;
...
-    int nbytes = size * sizeof( double );
-    double* ptr = (double*)fft_mem.smalloc( nbytes, mtype );
-    EXPECT_NE( ptr, nullptr );
-    fft_mem.sfree( ptr, mtype );
-}
 
 //---------------------------------------------------------------------------//
-void forwardReverseTest()
+template <class HostBackendType>
+void forwardReverseTest( bool use_default, bool use_params )
```
Comment (sfogerty) [inline cajita/unit_test/tstFastFourierTransform.hpp:36]: It strikes me as odd to have both `use_default` and `use_params` here - maybe I am missing something.
The way these are used, I might call them `use_host` and `use_params`, or something of that nature.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 17]
```diff
@@ -28,286 +28,548 @@ namespace Cajita
 namespace Experimental
```
Comment (sfogerty) [inline cajita/src/Cajita_FastFourierTransform.hpp:28]: I think we are nearing the point where these features could live outside the `Experimental` namespace. Should we take that step in this PR, or hold off?

**Comment by sslattery** [comment]
retest this please

File: cajita/unit_test/tstFastFourierTransform.hpp (hunk 16 above)
Comment (streeve) [reply cajita/unit_test/tstFastFourierTransform.hpp:36]: Fair point - what I meant was whether to use default backend and default parameters

File: core/unit_test/TestOPENMP_Category.hpp [hunk 18]
```diff
@@ -12,6 +12,7 @@
 #ifndef CABANA_TEST_OPENMP_CATEGORY_HPP
 #define CABANA_TEST_OPENMP_CATEGORY_HPP
 
+#define TEST_HOST
```
Comment (sslattery) [inline core/unit_test/TestOPENMP_Category.hpp:15]: This is not the place to put `TEST_HOST` - add something to the unit test that checks the backends and triggers the right type. These files are specific to a more general portable unit test strategy.

**Review by sslattery — CHANGES_REQUESTED** [review]

File: core/unit_test/TestOPENMP_Category.hpp (hunk 18 above)
Comment (streeve) [reply core/unit_test/TestOPENMP_Category.hpp:15]: I thought this was cleaner (thinking maybe this would come up in other places), but I can do this in the FFT unit test

**Comment by streeve** [comment]
Retest this please

File: cajita/unit_test/tstFastFourierTransform.hpp (hunk 9 above)
Comment (streeve) [reply cajita/unit_test/tstFastFourierTransform.hpp]: Replaced with a check for any enabled device backend

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 17 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp:28]: As we discussed, we'll hold off on this for now

File: core/unit_test/TestOPENMP_Category.hpp (hunk 18 above)
Comment (streeve) [reply core/unit_test/TestOPENMP_Category.hpp]: Removed

**Comment by streeve** [comment]
This is ready for (hopefully final) review

**Review by sslattery — APPROVED** [review]
